cedar-policy-validator/src/types.rs

```diff
@@ -904,9 +971,31 @@ impl EntityRecordKind {
         }
     }
 
+    /// Return `true` if this entity or record may have additional undeclared
+    /// attributes.
+    pub(crate) fn has_open_attributes_record(&self) -> bool {
+        match self {
+            // Records explicitly store this information.
+            EntityRecordKind::Record {
+                open_attributes, ..
+            } => open_attributes.is_open(),
+            // We know Actions never have additional attributes. This is true
+            // because the upper bound for any two action entities is
+            // `AnyEntity`, so if we have an ActionEntity here its attributes
+            // are known precisely.
+            EntityRecordKind::ActionEntity { .. } => false,
+            // The `AnyEntity` type has no declared attributes, but it is a
+            // super type of all other entity types which may have attributes,
+            // so it clearly may have additional attributes.
+            EntityRecordKind::AnyEntity => true,
+            // An entity LUB may not have additional attributes.
+            EntityRecordKind::Entity(_) => false,
```

Comment:
what if one of the components of the LUB has additional attributes?

Reply:
The types are organized a bit differently in Rust than in Dafny. In the Dafny it would have been difficult to let records be closed without letting entity types also have the open/closed flag, so entity types got the flag even though they don't really use it. For the Rust entity types don't have the flag.

Reply:
There's also a bit of subtlety because the record type returned by `get_attributes_type` might be open, but full entity LUB is still closed because it has access to all it's constituent entity types - which are closed - so it can account for every attribute that could possibly exist.

Reply:
Ack.  If it's subtle, it may be worth putting what you just wrote in a comment in the code here